### PR TITLE
Make DatasetFolder.find_classes public

### LINKED ISSUE #3627: Select subset of classes to sample from
## 🚀 Feature

When loading a dataset with `ImageFolder` provide an optional argument to select a subset of classes.

## Motivation

I deal with large (1000+) multi-class datasets upon which I train image classifiers. However, I usually don't want to train for all the classes at the same time.
## Pitch

I'd like to change `find_classes` function
https://github.com/pytorch/vision/blob/20a771e5143c6867eee63868c38a5bcc272a35e7/torchvision/datasets/folder.py#L61

to 

```python
classes = sorted(entry.name for entry in os.scandir(directory) 
                 if entry.is_dir() and (entry.name in allowed_classes or not allowed_classes))
```

where `allowed_classes: Optional[str] = []` is an empty list by default but it can given to `ImageFolder` at initialisation time (it has to be propagated back to `DatasetFolder` where `find_classes` is used).

## Alternatives

I tried to

1) manually create a new folder structure with only the relevant classes. This gets messy quite fast as I have now multiple versions of the same folders
2) resort to a custom dataloader filtering loaded samples after initialisation (obviously this is very slow as soon as the number of images increases)

## Additional context

Here I found a discussion on the same topic.

https://discuss.pytorch.org/t/how-to-sample-images-belonging-to-particular-classes/43776/8

cc @pmeier

### PR DESCRIPTION
**Pull request by alemelis**
Closes #3627

### CODE REVIEW

File: torchvision/datasets/folder.py [hunk 1]
```diff
@@ -203,7 +203,7 @@ def make_dataset(
         return make_dataset(directory, class_to_idx, extensions=extensions, is_valid_file=is_valid_file)
 
     @staticmethod
-    def _find_classes(dir: str) -> Tuple[List[str], Dict[str, int]]:
+    def find_classes(dir: str) -> Tuple[List[str], Dict[str, int]]:
```
Comment (NicolasHug) [inline torchvision/datasets/folder.py:206]: Let's make this an instance method too and remove the `@staticmethod` following https://github.com/pytorch/vision/pull/3496#pullrequestreview-619539316

Comment (NicolasHug) [reply torchvision/datasets/folder.py:206]: Also, we'll need docs now :)

**Review by NicolasHug — COMMENTED** [review]
Thanks for the PR @alemelis

File: torchvision/datasets/folder.py (hunk 1 above)
Comment (alemelis) [reply torchvision/datasets/folder.py:206]: yep instance method makes sense

there's already a quite detailed doctstring in `find_classes()` function, shall I move it to this method instead?

Comment (NicolasHug) [reply torchvision/datasets/folder.py:206]: I would link to `find_classes()` and  write something like

> Same as :func:`find_classes`. This method can be overridden to only consider a subset of classes, or to adapt to a different dataset directory structure.

File: torchvision/datasets/folder.py [hunk 2]
```diff
@@ -202,8 +202,11 @@ def make_dataset(
     ) -> List[Tuple[str, int]]:
         return make_dataset(directory, class_to_idx, extensions=extensions, is_valid_file=is_valid_file)
 
-    @staticmethod
-    def _find_classes(dir: str) -> Tuple[List[str], Dict[str, int]]:
+    def find_classes(self, dir: str) -> Tuple[List[str], Dict[str, int]]:
+        """
+        Same as :func:find_classes. This method can be overridden to only consider
```
Comment (NicolasHug) [inline torchvision/datasets/folder.py:207]: backticks + docstring format :)

```suggestion
        """Same as :func:`find_classes`.
        
        This method can be overridden to only consider 
```

**Review by NicolasHug — APPROVED** [review]
Nit but LGTM, thanks!

**Comment by NicolasHug** [comment]
@alemelis before we merge would you mind looking into the linting and  test issues?

**Comment by alemelis** [comment]
> @alemelis before we merge would you mind looking into the linting and test issues?

`torchhub_test` is failing with

```
ValueError: Cannot find master in https://github.com/pytorch/vision. If it's a commit from a forked repo, please call hub.load() with forked repo directly.
```

Since I'm pushing from a fork, is there an additional step I should follow to fix this?

**Comment by NicolasHug** [comment]
Thanks @alemelis !

**Comment by fmassa** [comment]
Thanks @alemelis for the PR and @NicolasHug for the review!

**Pull request merged**